server/src/main/java/org/elasticsearch/script/Metadata.java

```diff
@@ -55,26 +55,48 @@ public class Metadata {
     public static FieldProperty<Number> LongField = new FieldProperty<>(Number.class).withValidation(FieldProperty.LONGABLE_NUMBER);
 
     protected final Map<String, Object> map;
-    protected final Map<String, FieldProperty<?>> properties;
+    private final Map<String, FieldProperty<?>> properties;
     protected static final FieldProperty<?> BAD_KEY = new FieldProperty<>(null, false, false, null);
 
-    public Metadata(Map<String, Object> map, Map<String, FieldProperty<?>> properties) {
+    /**
+     * Constructs a new Metadata object represented by the given map and properties.
+     * <p>
+     * The passed-in map is used directly -- subsequent modifications to it outside the methods of this class may result in
+     * undefined behavior. By contrast, the properties map is copied via {@link Map#copyOf(Map)}) in order to ensure that it cannot be
+     * changed.
+     * <p>
+     * As an implementation detail, since {@link Map#copyOf(Map)} is a no-op if the passed-in instance is an "unmodifiable map"
+     * (as defined by {@link Map#copyOf(Map)}), callers of this constructor should construct their passed-in properties map via
+     * {@link Map#of(Object, Object)}} in order to ensure optimal performance.
```

Comment:
Can we check this using an assertion?

Reply:
Sure thing! That looks like 81073bd60fecfd05997a167366615be46bdaae87. I didn't go down that road initially because I thought perhaps it might make the tests annoying to write, but in all actuality it's not that bad. What do you think?

Reply:
Looks sensible